module/zfs/metaslab.c

```diff
@@ -1255,9 +1255,10 @@ metaslab_init(metaslab_group_t *mg, uint64_t id, uint64_t object, uint64_t txg,
 
 	/*
 	 * We only open space map objects that already exist. All others
-	 * will be opened when we finally allocate an object for it.
+	 * will be opened when we finally allocate an object for it.  For
+	 * readonly pools there is no need to open the space map object.
 	 */
-	if (object != 0) {
+	if (object != 0 && spa_writeable(vd->vdev_spa)) {
```

Comment:
Are we sure that *spa_writeable(vd->vdev_spa)* returns true in all contexts here? I'm getting the suspicion that #3094 is connected to this or #2909, and this line may explain the pool importing itself sans space map and corrupting all future space maps by writing data during the period it is imported this way without updating the space maps.